events/importer/espoo.py

```diff
@@ -1,800 +1,534 @@
-    "kulttuuri ja liikunta  ",
-    "kaikki tapahtumat",
-    "muut tapahtumat",
-    "sellosali",
-    "espoon kulttuurikeskus",
-    "espoon kaupunginmuseo",
-    "kamu",
-    "näyttelykeskus weegee",
-    "karatalo",
-    "ohjelmisto",
-    "kulttuurikohteet ja -toimijat",
-    "espoo.fi",
-    "kulttuuriespoo.fi",
-    "kulttuurikeskukset ja -talot",
-]
-
-LOCATIONS = {
-    # Place name in Finnish -> ((place node ids in event feed), tprek id)
-    "matinkylän asukaspuisto": ((15728,), 20267),
-    "soukan asukaspuisto": ((15740,), 20355),
-    "espoon kulttuurikeskus": ((15325,), 58548),
-    "näyttelykeskus weegee": ((15349,), 20404),
-    "KAMU": ((28944,), 20405),
-    "Karatalo": ((15357,), 21432),
-    "Nuuksio": ((15041,), 28401),
-    "Olarin asukaspuisto": ((15730,), 20268),
-    "Lasten kulttuurikeskus Aurora": ((15350,), 21431),
-    "Suviniityn avoin päiväkoti": ((15781,), 20376),
-    "Sellosali": ((15281,), 59212),
-    "Talomuseo Glims": ((28954,), 59312),
-}
-
-ESPOO_BASE_URL = "http://www.espoo.fi"
-ESPOO_API_URL = (
-    ESPOO_BASE_URL + "/api/opennc/v1/ContentLanguages({lang_code})"
-    "/Contents?$filter=TemplateId eq 58&$expand=ExtendedProperties,LanguageVersions"
-    "&$orderby=EventEndDate desc&$format=json"
-)
-
-ESPOO_LANGUAGES = {
-    "fi": 1,
-    "sv": 3,
-    "en": 2,
-}
-
-ADDRESS_LANGUAGES = ("fi", "sv")
-
-LOCAL_TZ = timezone("Europe/Helsinki")
-
-
-def get_lang(lang_id):
-    for code, lid in ESPOO_LANGUAGES.items():
-        if lid == lang_id:
-            return code
-    return None
-
-
-def mark_deleted(obj):
-    if obj.deleted:
-        return False
-    obj.deleted = True
-    obj.save(update_fields=["deleted"])
-    return True
-
-
-def clean_street_address(address):
-    latin1_charset = "a-zàáâãäåæçèéêëìíîïðñòóôõö÷øùúûüýþÿ"
-
-    address = address.strip()
-    pattern = re.compile(
-        r"([%s\ -]*[0-9-\ ]*\ ?[a-z]{0,2}),?\ *(0?2[0-9]{3})?\ *(espoo|esbo)?"
-        % latin1_charset,
-        re.I,
-    )
-    match = pattern.match(address)
-    if not match:
-        logger.warning("Address not matching {}".format(address))
-        return {}
-    groups = match.groups()
-    street_address = groups[0]
-    postal_code = None
-    city = None
-    if len(groups) == 2:
-        city = groups[1]
-    elif len(groups) == 3:
-        postal_code = groups[1]
-        city = groups[2]
+
+def parse_jsonld_id(data: dict) -> str:
+    ld_reference = data.get("@id")
+    if ld_reference:
+        origin_id = ld_reference.split("/")[-2]
+        if origin_id:
+            return origin_id
+    raise RuntimeError("Could not parse jsonld id")
+
+
+def parse_jsonld_ids(data: list) -> list[str]:
+    return [parse_jsonld_id(d) for d in data]
+
+
+def _get_data(url: str) -> dict:
+    for _ in range(settings.ESPOO_MAX_RETRIES):
+        response = requests.get(url)
+        if response.status_code == 200:
+            return response.json()
+        time.sleep(settings.ESPOO_WAIT_BETWEEN)
```

Comment:
Not supported by requests_mock ( https://github.com/jamielennox/requests-mock/issues/223 )

Reply:
This hurts 🥲 test code should never stop you from implementing stuff in production code. Tests are there to validate that your code works as intended. If you use `HTTPAdapter` & `Retry`, you shouldn't need to test the retrying. They're presumably tested in `requests`/`urllib3` already. You could test for `max_retries` or something too if you want to be extra sure with having those retries in place.

That being said, this change is not exactly needed, it's just if you want to use something existing already. `Retry` does include handy stuff like backoffs and retrying only for certain status codes. A custom implementation is fine too, especially since it's a rather limited and simple use case, so do what you will.